spatialos-sdk/src/worker/schema/primitives.rs

```diff
@@ -56,29 +77,54 @@ macro_rules! impl_primitive_field {
                 }
             }
 
-            fn count(object: &SchemaObject, field: FieldId) -> usize {
-                unsafe { $schema_count(object.as_ptr(), field) as usize }
-            }
+            fn add_list(object: &mut SchemaObject, field: FieldId, data: &[$rust_type]) {
+                // Determine how large the buffer needs to be.
+                //
+                // NOTE: We allocate extra padding when allocating the buffer in order to have room
+                // to adjust the alignment of the buffer to match the alignment of `Self::RustType`
+                // and still have enough room in the buffer for the right number of elements.
+                let byte_len = {
+                    let data_len = data.len() * mem::size_of::<Self::RustType>();
+                    let padding = mem::align_of::<Self::RustType>() - 1;
```

Comment:
The idea is that if we can't assume that the buffer returned by `Schema_AllocateBuffer` is properly aligned for our type, we may need to "shift" the buffer by up to `alignment - 1` bytes in order to get proper alignment. We therefore need to allocate an extra `alignment - 1` bytes of padding in order to ensure we have enough space in the buffer to copy all of the data after we've fixed the alignment.

... as I'm writing this, it occurs to me that this is *probably* not the right thing to do. For example, if we leave 3 bytes of padding at the beginning of the buffer in order to get proper alignment, there's probably going to be 3 bytes of garbage at the beginning of the buffer when the data is received by another worker. Any worker trying to read this data would have to somehow know how much padding was added. We could try to encode that in an extra byte at the beginning of the buffer, but that really feels like something I'd expect to either be documented as a necessary step or be handled by the SDK directly. @dgavedissian do you know if there are any alignment guarantees around `Schema_AllocateBuffer`? I feel like I've asked this before, and I wouldn't be surprised if it guaranteed large enough alignment to be valid for all primitive types.

Reply:
>  For example, if we leave 3 bytes of padding at the beginning of the buffer in order to get proper alignment, there's probably going to be 3 bytes of garbage at the beginning of the buffer when the data is received by another worker.

I'm not sure this is necessarily true, since we pass in the pointer to the aligned buffer to the `Schema_AddxxxList` method. So the garbage bytes shouldn't ever be read by the schema API, since the buffer isn't directly sent over the wire, it's just a allocation that has its lifetime tied to the schema object.

Reply:
Huh, that's a good point. So maybe its fine after all? Is there a good example of how this is handled in the SDKs for other languages? That would be a good way to confirm how to correctly handle this, I think.

Reply:
Pasting the two implementations side-by-side

**C#**
```csharp
        /// <summary>
        /// Functions that append a list of primitive values for a particular field ID to an object. Note
        /// that, for best performance, fields should be added to the object in field ID order.
        /// This function will allocate a copy of the supplied <c>values</c> parameter.
        /// </summary>
        public void AddFloatList(FieldId fieldId, float[] values)
        {
            Uint32 bufferLength = (Uint32) (values.Length * Marshal.SizeOf(typeof(float)));
            Uint8* buffer = AllocateBuffer(bufferLength);
            fixed (float* valueBuffer = values)
            {
                CWorker.Memcpy(buffer, (Uint8*) valueBuffer, (UIntPtr) bufferLength);
            }

            AddFloatList(fieldId, (float*) buffer, values.Length);
        }
```

**Rust**

```rust
            fn add_list(object: &mut SchemaObject, field: FieldId, data: &[$rust_type]) {
                // Determine how large the buffer needs to be.
                //
                // NOTE: We allocate extra padding when allocating the buffer in order to have room
                // to adjust the alignment of the buffer to match the alignment of `Self::RustType`
                // and still have enough room in the buffer for the right number of elements.
                let byte_len = {
                    let data_len = data.len() * mem::size_of::<Self::RustType>();
                    let padding = mem::align_of::<Self::RustType>() - 1;
                    data_len + padding
                };

                // Allocate a buffer that is owned by `object`.
                let buffer = unsafe {
                    let data_ptr = Schema_AllocateBuffer(object.as_ptr_mut(), byte_len as _);
                    let buffer = slice::from_raw_parts_mut(data_ptr, byte_len);

                    // Convert the byte buffer into a correctly-aligned slice of the data type.
                    let (_prefix, buffer, _suffix) = buffer.align_to_mut::<Self::RustType>();
                    buffer
                };

                // Populate the buffer.
                buffer.copy_from_slice(data);

                unsafe {
                    $schema_add_list(
                        object.as_ptr_mut(),
                        field,
                        buffer.as_ptr() as *const _,
                        buffer.len() as u32,
                    );
                }
            }
```
Two things I noticed: 

* There is no alignment in the C# version. That difference may be that we directly copy the memory across (`CWorker.Memcpy` aliases the C `memcpy` function). 
* `from_raw_parts_mut` requires `data_ptr` to be aligned in the contract and we do the alignment _after_ we call this. 

---

Had a quick peek at the implementation and the memory is allocated with `std::max_align_t` as the alignment (https://en.cppreference.com/w/cpp/types/max_align_t) which means we shouldn't need to worry about the alignment if I'm reading the cpp docs correctly: 

> std::max_align_t is a trivial standard-layout type whose alignment requirement is at least as strict (as large) as that of every scalar type.

Reply:
Ah! I swear one of you two previously told me that it uses `max_align_t` under the hood 🤦‍♀. Thinking about it, though, we could probably do a similar copy in the Rust version. Currently we convert the raw buffer into a mutable slice of the appropriate type, and then copy over to that slice. Instead, we could convert the input data into a byte slice and copy *that* into the allocated buffer. I think doing it that way would be both simpler and would keep the Rust version more consistent with the other SDKs.

Reply:
Eh, okay I say that *but* it's actually much easier to convert the byte buffer to a typed slice than to go the other way around 🤔 So I'm actually going just remove the alignment logic and leave a comment noting that it's safe because the C SDK guarantees max alignment.